rest-service/manager_rest/resource_manager.py

```diff
@@ -1057,6 +1061,76 @@ def add_node_instance_from_dict(self, instance_dict):
         # adding it to the response.
         instance_dict['resource_availability'] = resource_availability
 
+    def list_agents(self, deployment_id=None, node_ids=None,
+                    node_instance_ids=None, install_method=None):
+        filters = {}
+        if deployment_id is not None:
+            filters['deployment_id'] = deployment_id
+        if node_ids is not None:
+            filters['id'] = node_ids
+        nodes = {n._storage_id: n
+                 for n in self.sm.list(models.Node, filters=filters,
+                                       get_all_results=True)
+                 if self._is_agent_node(n)}
+
+        if not nodes:
+            pagination = {'total': 0, 'size': 0, 'offset': 0}
+            return ListResult([], metadata={'pagination': pagination})
+
+        instance_filters = {'_node_fk': list(nodes)}
+        if node_instance_ids:
+            instance_filters['id'] = node_instance_ids
+        node_instances = self.sm.list(models.NodeInstance,
+                                      filters=instance_filters,
+                                      get_all_results=True)
+        agents = []
+        for inst in node_instances:
+            agent = self._agent_from_instance(inst)
+            if agent is None:
+                continue
+            if install_method is not None and \
+                    agent['install_method'] not in install_method:
+                continue
+            agents.append(agent)
+
+        # there's no real way to implement pagination in a meaningful manner,
+        # since we need to query node-instances anyway; still, include the
+        # sizes to keep response object shape consistent with all other
+        # endpoints
+        pagination = {'total': len(agents), 'size': len(agents), 'offset': 0}
+        return ListResult(agents, metadata={'pagination': pagination})
+
+    def _agent_from_instance(self, instance):
+        if instance.state != 'started':
+            return
+        agent = instance.runtime_properties.get('cloudify_agent')
+        if not agent:
+            return
+        if agent.get('windows'):
+            system = 'windows'
+        else:
+            system = agent.get('distro')
+            if agent.get('distro_codename'):
+                system = '{0} {1}'.format(system, agent.get('distro_codename'))
```

Comment:
Can (or should) this also run some sort of check of agent health?

Reply:
no, we have `cfy agents validate` for that

ideally i'd like to have it here as well, but it's unwieldy, since it would require querying the agents, which should be async and not in the restservice... so can't really do that right now